### ISSUE TITLE
[dev.icinga.com #12941] Append arguments in command templates/objects

### ISSUE THREAD

**Issue by icinga-migration** [Icinga/icingaweb2-module-director#499]
This issue has been migrated from Redmine: https://dev.icinga.com/issues/12941

**Created by leeclemens on 2016-10-18 17:14:51 +00:00**

Assignee: _tgelf_
Status: _Resolved_  (closed on _2016-11-15 17:00:03 +00:00_)
Target Version: _1.3.0_
Last Update: _2016-11-15 17:00:03 +00:00 (in Redmine)_

---

I specifically used NotificationCommand templates and an object, but the limitation may apply to other Command types.

When creating multiple templates, the arguments are replaced (as Markus Frosch pointed out).
There doesn't appear to be a way in Director to append them, although they import each other properly.

Example:

    template NotificationCommand "Email Template" {
        import "plugin-notification-command"
        arguments = {
            "--host-address" = {
                required = true
                value = "$host.address$"
            }
            "--host-display-name" = {
                required = true
                value = "$host.display_name$"
            }
            "--host-name" = {
                required = true
                value = "$host.name$"
            }
            "--notification-author" = {
                required = true
                value = "$notification.author$"
            }
            "--notification-comment" = {
                required = true
                value = "$notification.comment$"
            }
            "--notification-type" = {
                required = true
                value = "$notification.type$"
            }
            "--user-email" = {
                required = true
                value = "$user.email$"
            }
        }
    }

    template NotificationCommand "Email Host Template" {
        import "plugin-notification-command"
        import "Email Template"

        command = [ "/etc/icinga2/scripts/mail-host.py" ]
        arguments = {
            "--host-output" = {
                required = true
                value = "$host.output$"
            }
            "--host-state" = {
                required = true
                value = "$host.state$"
            }
        }
    }

    object NotificationCommand "Standard Email Host" {
        import "plugin-notification-command"
        import "Email Host Template"

    }

    [2016-10-17 12:38:02 -0400] warning/PluginNotificationTask: Notification command for object 'hostname.example.com' (PID: 8596, arguments: '/etc/icinga2/scripts/mail-host.py' '--host-output' 'PING CRITICAL - Packet loss = 100%' '--host-state' 'DOWN') terminated with exit code 2, output: usage: mail-host.py [-h] --host-name HOST_NAME --host-address HOST_ADDRESS 

Markus' response:

    What you should do: ( += )

        arguments += {
            "--host-output" = {
                required = true
                value = "$host.output$"
            }
            "--host-state" = {
                required = true
                value = "$host.state$"
            }
        }

It may be safe to always use += instead of = ? But I haven't tested it (or there could be other issues caused).
Maybe a new option on the Argument tab(s) to "Append"?

References:
https://lists.icinga.org/pipermail/icinga-users/2016-October/011413.html
https://lists.icinga.org/pipermail/icinga-users/2016-October/011415.html

**Changesets**

_2016-11-15 16:57:02 +00:00_ by _tgelf_ d5131f7fac442f37b2d5b9428887e95365830d27

    Arguments: always render += instead of =
    
    This allows to extend existing arguments
    
    fixes #12941

---

**Relations**:

* relates [#12941](https://dev.icinga.com/issues/12941)
